corehq/apps/app_manager/app_translations/app_translations.py

```diff
@@ -49,7 +52,26 @@ def none_or_unicode(val):
     return rows
 
 
-def process_bulk_app_translation_upload(app, f):
+def read_uploaded_app_translation_file(f, msgs):
+    try:
+        workbook = WorkbookJSONReader(f)
+    # todo: HeaderValueError does not belong here
+    except (HeaderValueError, InvalidExcelFileException) as e:
+        msgs.append(
+            (messages.error, _(APP_TRANSLATION_UPLOAD_FAIL_MESSAGE).format(e))
+        )
+        return False
+    except JSONReaderError as e:
+        msgs.append(
+            (messages.error, _(
+                "App Translation Failed! There is an issue with excel columns. Error details: {}."
+            ).format(e))
+        )
+        return False
+    return workbook
```

Comment:
Not a criticism, just curious if there is an underlying principle behind your choice here: You've chosen to return `workbook`, but pass `msgs` in by reference, and update it as a side-effect, but not read from it. The more obvious would be ...
```python
def read_uploaded_app_translation_file(f):
    msgs = []
    # ... etc. ...
    return workbook, msgs
```
... so I think you probably considered that and intentionally chose the approach you took. 

I'm not suggesting you change your code to this. I just want to learn why sometimes one would want to return some values explicitly, and other values by reference (even when you don't want to read the incoming value).

Reply:
So in this case it would be 
```
def read_uploaded_app_translation_file(f):
    msgs = []
    # ... etc. ...
    return workbook, msgs
```
and then where this is called
```
    workbook, msgs = read_uploaded_app_translation_file(f)
```

the only advantage in the current implementation i see is that we have `msgs` declared only once whereas in the code snippet above we have it at two places and there is no need for them to keep separate variables. 

I guess the current implementation would make more sense if `msgs` was not an empty string but with some messages already added earlier before its passed to `read_uploaded_app_translation_file` and we would then modify the same if needed.
If the change makes things more clear, happy to change it. Passing by reference has confused me at times as well but i guess that is just a habit thing.

Reply:
Yeah, I agree that changing values passed by reference can be confusing, and harder to see what's happening. (But sometimes it's very useful.)

I think what you're doing here is obvious and not confusing at all, so happy happy to keep this.

Reply:
updated to return msgs from this method.